File: ui-service/src/lib/workflow/kycStepIntakeCatalog.ts
export interface KycStepIntakeMeta {
  fieldKey: string | null
  defaultDocumentTypes: string[]
  label: string
}

/** Human-readable labels for workflow-resolved intake document slots. */
export const INTAKE_DOCUMENT_TYPE_LABELS: Record<string, string> = {
  PAN_CARD: 'PAN card',
  AADHAAR: 'Aadhaar',
  VOTER_ID: 'Voter ID card',
  DRIVING_LICENSE: 'Driving licence',
  GST_RETURN: 'GST return (GSTR)',
  BANK_STATEMENT: 'Bank statement',
  PHOTOGRAPH: 'Photograph',
}

export const KYC_STEP_INTAKE_CATALOG: Record<string, KycStepIntakeMeta> = {
  PAN_VERIFY: { fieldKey: 'panNumber', defaultDocumentTypes: ['PAN_CARD'], label: 'PAN' },
  AADHAAR_OTP: { fieldKey: 'aadhaar', defaultDocumentTypes: ['AADHAAR'], label: 'Aadhaar' },
  VOTER_ID_VERIFY: { fieldKey: 'voterId', defaultDocumentTypes: ['VOTER_ID'], label: 'Voter ID' },
  DL_VERIFY: { fieldKey: 'dlNumber', defaultDocumentTypes: ['DRIVING_LICENSE'], label: 'Driving licence' },
  GSTIN_VERIFY: { fieldKey: 'gstin', defaultDocumentTypes: ['GST_RETURN'], label: 'GSTIN' },
  BANK_PENNY_DROP: { fieldKey: 'bankAccountNumber', defaultDocumentTypes: ['BANK_STATEMENT'], label: 'Bank account' },
  FACE_MATCH: { fieldKey: null, defaultDocumentTypes: ['PHOTOGRAPH'], label: 'Face match' },
  LIVENESS: { fieldKey: null, defaultDocumentTypes: ['PHOTOGRAPH'], label: 'Liveness' },
  UDYAM_VERIFY: { fieldKey: 'udyam', defaultDocumentTypes: [], label: 'Udyam' },
  CIN_MCA21: { fieldKey: 'cin', defaultDocumentTypes: [], label: 'CIN' },
}

export function metaForKycStep(step: string): KycStepIntakeMeta | null {
  return KYC_STEP_INTAKE_CATALOG[step.trim().toUpperCase()] ?? null
}

export function stepNameFromWorkflowStep(step: Record<string, unknown>): string {
  return String(step.step ?? '').trim().toUpperCase()
}

export function configuredKycStepNames(steps: Record<string, unknown>[]): string[] {
  return steps.map((s) => stepNameFromWorkflowStep(s)).filter(Boolean)
}

File: services/los-core-service/src/main/java/com/los/core/service/workflow/intake/KycStepIntakeCatalog.java
package com.los.core.service.workflow.intake;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Maps KYC workflow step types to intake field keys and default document types.
 */
public final class KycStepIntakeCatalog {

    public static final String POLICY_LEGACY = "LEGACY";
    public static final String POLICY_WORKFLOW_DRIVEN = "WORKFLOW_DRIVEN";

    private KycStepIntakeCatalog() {
    }

    public record StepIntakeMeta(String fieldKey, List<String> defaultDocumentTypes) {
    }

    private static final Map<String, StepIntakeMeta> CATALOG = Map.ofEntries(
            Map.entry("PAN_VERIFY", new StepIntakeMeta("panNumber", List.of("PAN_CARD"))),
            Map.entry("AADHAAR_OTP", new StepIntakeMeta("aadhaar", List.of("AADHAAR"))),
            Map.entry("VOTER_ID_VERIFY", new StepIntakeMeta("voterId", List.of("VOTER_ID"))),
            Map.entry("DL_VERIFY", new StepIntakeMeta("dlNumber", List.of("DRIVING_LICENSE"))),
            Map.entry("GSTIN_VERIFY", new StepIntakeMeta("gstin", List.of("GST_RETURN"))),
            Map.entry("BANK_PENNY_DROP", new StepIntakeMeta("bankAccountNumber", List.of("BANK_STATEMENT"))),
            Map.entry("FACE_MATCH", new StepIntakeMeta(null, List.of("PHOTOGRAPH"))),
            Map.entry("LIVENESS", new StepIntakeMeta(null, List.of("PHOTOGRAPH"))),
            Map.entry("UDYAM_VERIFY", new StepIntakeMeta("udyam", List.of())),
            Map.entry("CIN_MCA21", new StepIntakeMeta("cin", List.of()))
    );

    public static StepIntakeMeta metaForStep(String stepName) {
        if (stepName == null || stepName.isBlank()) {
            return null;
        }
        return CATALOG.get(stepName.trim().toUpperCase());
    }

    public static boolean isWorkflowDriven(Map<String, Object> intakeConfig) {
        if (intakeConfig == null || intakeConfig.isEmpty()) {
            return false;
        }
        return POLICY_WORKFLOW_DRIVEN.equalsIgnoreCase(String.valueOf(intakeConfig.getOrDefault("policy", "")));
    }

    public static Map<String, Object> defaultWorkflowDrivenIntakeConfig() {
        Map<String, Object> personalFields = new LinkedHashMap<>();
        personalFields.put("dateOfBirth", Map.of("collect", true, "required", true));
        personalFields.put("gender", Map.of(
                "collect", true,
                "required", false,
                "allowedValues", List.of("MALE", "FEMALE", "OTHER", "PREFER_NOT_TO_SAY")));

        Map<String, Object> config = new LinkedHashMap<>();
        config.put("policy", POLICY_WORKFLOW_DRIVEN);
        config.put("personalFields", personalFields);
        config.put("ageRules", Map.of("enabled", false, "minAge", 18, "maxAge", 70));
        config.put("tenureRules", Map.of("inputMode", "numeric", "min", 1, "max", 360));
        config.put("mandatoryFieldGroups", List.of());
        config.put("standaloneDocuments", List.of());
        return config;
    }

    public static Map<String, Object> legacyIntakeConfig() {
        return Map.of("policy", POLICY_LEGACY);
    }

    /**
     * Returns true when the merged KYC payload has the minimum identity input to attempt this step.
     */
    public static boolean hasExecutionPayload(String stepName, Map<String, Object> payload) {
        if (stepName == null || stepName.isBlank()) {
            return false;
        }
        String step = stepName.trim().toUpperCase();
        return switch (step) {
            case "PAN_VERIFY" -> hasAny(payload, "panNumber", "pan");
            case "AADHAAR_OTP" -> hasAny(payload, "aadhaarNumber", "aadhaarLast4", "aadhaar");
            case "VOTER_ID_VERIFY" -> hasAny(payload, "epicNo", "voterId");
            case "DL_VERIFY" -> hasAny(payload, "dlNo", "drivingLicenseNumber", "dlNumber");
            case "GSTIN_VERIFY" -> hasAny(payload, "gstin");
            case "UDYAM_VERIFY" -> hasAny(payload, "udyamRegistrationNo", "udyam", "udyamNumber");
            case "BANK_PENNY_DROP" -> hasAny(payload, "accountNumber", "bankAccountNumber")
                    && hasAny(payload, "ifsc");
            case "CIN_MCA21" -> hasAny(payload, "cin");
            case "MNRL" -> hasAny(payload, "mobile", "phone");
            default -> true;
        };
    }

    private static boolean hasAny(Map<String, Object> payload, String... keys) {
        if (payload == null) {
            return false;
        }
        for (String key : keys) {
            Object value = payload.get(key);
            if (value != null && !String.valueOf(value).trim().isEmpty()) {
                return true;
            }
        }
        return false;
    }
}
